compiler/optimizer/parallelize.go

```diff
@@ -2,189 +2,211 @@ package optimizer
 func orderAsDirection(which order.Which) int {
 	if which == order.Asc {
 		return 1
 	}
 	return -1
 }
 
-// XXX assume the trunk is from a from op at seq.Ops[0] and we will
-// possible insert an operator at seq.Op[1]
-func (o *Optimizer) parallelizeTrunk(seq *dag.Sequential, trunk *dag.Trunk, replicas int) error {
-	from, ok := seq.Ops[0].(*dag.From)
-	if !ok {
-		return errors.New("internal error: parallelizeTrunk: entry is not a From")
+func (o *Optimizer) parallelizeScan(ops []dag.Op, replicas int) ([]dag.Op, error) {
+	// For now we parallelize only pool scans and no metadata scans.
+	// We can do the latter when we want to scale the performance of metadata.
+	if replicas < 2 {
+		return nil, fmt.Errorf("internal error: parallelizeScan: bad replicas: %d", replicas)
 	}
-	if len(from.Trunks) > 1 {
-		// No support for multi-trunk from's yet, which only arise for
-		// joins and other peculaliar mixins of different sources.
-		// We need to handle join parallelization differently, though
-		// the logic is not very far from what's here.
-		return nil
+	if len(ops) < 1 {
+		return nil, errors.New("internal error: parallelizeScan: short path")
 	}
-	if len(from.Trunks) == 0 {
-		return errors.New("internal error: no trunks in dag.From")
+	scan, ok := ops[0].(*dag.SeqScan)
+	if !ok {
+		return nil, errors.New("internal error: parallelizeScan: entry has no dag.SeqScan")
 	}
-	egressSortKey, err := o.sortKeyOfTrunk(trunk)
+	srcSortKey, err := o.sortKeyOfSource(scan)
 	if err != nil {
-		return err
+		return nil, err
 	}
-	if len(egressSortKey.Keys) > 1 {
-		// XXX don't yet support multi-key ordering
-		return nil
+	if len(srcSortKey.Keys) > 1 {
+		// XXX Don't yet support multi-key ordering.  See Issue #2657.
+		return nil, nil
 	}
-	// This logic requires that there is only one trunk in the From,
-	// as checked above.
-	sortKey, err := o.sortKeyOfSource(trunk.Source, order.Nil)
-	if err != nil {
-		return err
+	if len(ops) == 1 && scan.Filter == nil {
+		return nil, nil
 	}
-	inlineSequentials(seq)
-	// Check that the path consisting of the original from
+	// concurrentPath will check that the path consisting of the original source
 	// sequence and any lifted sequence is still parallelizable.
-	if trunk.Seq != nil && len(trunk.Seq.Ops) > 0 {
-		n, newSortKey, err := o.splittablePath(trunk.Seq.Ops, sortKey)
-		if err != nil {
-			return err
-		}
-		if n != len(trunk.Seq.Ops) {
-			return nil
+	n, outputKey, _, needMerge, err := o.concurrentPath(ops[1:], srcSortKey)
+	if err != nil {
+		return nil, err
+	}
+	// XXX Fix this to handle multi-key merge. See Issue #2657.
+	if len(outputKey.Keys) > 1 {
+		return nil, nil
+	}
+	head := ops[:n+1]
+	tail := ops[n+1:]
+	par := &dag.Parallel{Kind: "Parallel", Any: true}
+	for k := 0; k < replicas; k++ {
+		par.Ops = append(par.Ops, &dag.Sequential{Kind: "Sequential", Ops: copyOps(head)})
+	}
+	var merge dag.Op
+	if !needMerge {
```

Comment:
Nit: I think this would be a little clearer if you reversed the sense here (and swapped the "if" and "else" blocks).
```suggestion
	if needMerge {
```